node/src/services/blockchain/transaction.rs

```diff
@@ -0,0 +1,150 @@
+
+use crate::services::blockchain::{commands::BlockchainServiceInterface, types::ExtrinsicResult};
+
+use super::{types::ExtrinsicHash, BlockchainService};
+
+const LOG_TARGET: &str = "blockchain-transaction";
+
+#[derive(Debug)]
+pub struct SubmittedTransaction {
+    watcher: Receiver<String>,
+    hash: ExtrinsicHash,
+    timeout: Option<Duration>,
+}
+
+const NO_TIMEOUT_WARN: Duration = Duration::from_secs(60);
+
+impl SubmittedTransaction {
+    pub fn new(watcher: Receiver<String>, hash: H256) -> Self {
+        Self {
+            watcher,
+            hash,
+            timeout: None,
+        }
+    }
+
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    pub async fn watch_for_success(
+        &mut self,
+        blockchain: &ActorHandle<BlockchainService>,
+    ) -> anyhow::Result<()> {
+        // Wait for the transaction to be included in a block.
+        let block_hash;
+
+        let start_time = Instant::now();
+        loop {
+            // Get the elapsed time since submit.
+            let elapsed = start_time.elapsed();
+            // Calculate the remaining time to wait.
+            let remaining = match self.timeout {
+                Some(timeout) => {
+                    // Check if the timeout has been reached.
+                    if elapsed > timeout {
+                        return Err(anyhow!(
+                            "Timeout waiting for transaction to be included in a block"
+                        ));
+                    }
+
+                    timeout - elapsed
+                }
+                None => NO_TIMEOUT_WARN,
+            };
+
+            // Wait for either a new message from the watcher, or the timeout to be reached.
+            let result = match tokio::time::timeout(remaining, self.watcher.recv()).await {
+                Ok(result) => match result {
+                    Some(result) => result,
+                    None => {
+                        return Err(anyhow!("Transaction watcher channel closed"));
+                    }
+                },
+                Err(_) => {
+                    // Timeout reached, exit the loop.
+                    match self.timeout {
+                        Some(_) => {
+                            return Err(anyhow!(
+                                "Timeout waiting for transaction to be included in a block"
+                            ));
+                        }
+                        None => {
+                            // No timeout set, continue waiting.
+                            warn!(target: LOG_TARGET, "No timeout set and {:?} elapsed, continuing to wait for transaction to be included in a block.", NO_TIMEOUT_WARN);
+
+                            continue;
+                        }
+                    }
+                }
+            };
+            // Parse the JSONRPC string, now that we know it is not an error.
```

Comment:
This comment was left here, and now it is a bit misleading. 
```suggestion
            // Parse the JSONRPC string. The strings sent by the RPC wacher should be valid JSONRPC strings.
```